csp/utils.py

```diff
@@ -178,3 +275,94 @@ def build_script_tag(content=None, **kwargs):
     c = _unwrap_script(content) if content and not kwargs.get('src') else ''
     attrs = ATTR_FORMAT_STR.format(**data).rstrip()
     return ('<script{}>{}</script>'.format(attrs, c).strip())
+
+
+def kwarg_to_directive(kwarg, value=None):
+    return setting_to_directive(kwarg, prefix='', value=value)
+
+
+def _clean_input_policy(policy):
+    return dict(
+        kwarg_to_directive(in_directive, value=value)
+        for in_directive, value in policy.items()
+    )
+
+
+def csp_definitions_update(csp_definitions, other):
+    """ Update one csp defnitions dictionary with another """
+    if isinstance(other, dict):
+        other = other.items()
+    for name, csp in other:
+        csp_definitions.setdefault(name, {}).update(csp)
+    return csp_definitions
+
+
+class PolicyNames:
+    length = 20
+    last_policy_name = None
+
+    def __next__(self):
+        self.last_policy_name = get_random_string(self.length)
+        return self.last_policy_name
+
+    def __iter__(self):
+        return self
+policy_names = PolicyNames()
+
+
+def iter_policies(policies):
+    """
+    Accepts the following formats:
+    - a policy dictionary (formatted like in settings.CSP_POLICY_DEFINITIONS)
+    - an iterable of policies: (item, item, item,...)
+
+    item can be any of the following:
+        - string representing a named policy in settings.CSP_POLICY_DEFINITIONS
+        - subscriptable two-tuple: (name, csp)
+        - csp dictionary which will be assigned a random name
+
+    Yields a tuple: (name, csp)
+    """
+    if isinstance(policies, dict):
+        yield from (
+            (name, _clean_input_policy(policy))
+            for name, policy in policies.items()
+        )
+        return
+
+    for policy in policies:
+        if isinstance(policy, str):
+            # Named policies will be handled by the middleware
+            yield (policy, {})
+        elif isinstance(policy, (list, tuple)):
+            yield (policy[0], _clean_input_policy(policy[1]))
+        else:  # dictionary containing a single csp
+            yield (next(policy_names), _clean_input_policy(policy))
+
+
+def _policies_from_names_and_kwargs(
+    csp_names=('default',),
+    csp_definitions=None,
+    **kwargs,
+):
+    """
+    Helper used in csp_update and csp_replace to process args
+    """
+    if csp_definitions:
+        csp_definitions = csp_definitions_update(kwargs, csp_definitions)
+
+    if csp_names is None:
+        update = csp_definitions
+
+    if not kwargs:
+        update = OrderedDict(
```

Comment:
_policies_from_names_and_kwargs: This seems wrong in the case where definitions are passed in.